src/hotspot/share/cds/aotLinkedClassBulkLoader.cpp

```diff
@@ -0,0 +1,204 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "cds/aotLinkedClassBulkLoader.hpp"
+#include "cds/aotLinkedClassTable.hpp"
+#include "cds/archiveBuilder.hpp"
+#include "cds/archiveUtils.inline.hpp"
+#include "cds/cdsConfig.hpp"
+#include "cds/cdsProtectionDomain.hpp"
+#include "cds/heapShared.hpp"
+#include "cds/lambdaFormInvokers.inline.hpp"
+#include "classfile/classLoader.hpp"
+#include "classfile/classLoaderDataGraph.hpp"
+#include "classfile/classLoaderExt.hpp"
+#include "classfile/dictionary.hpp"
+#include "classfile/javaClasses.hpp"
+#include "classfile/systemDictionary.hpp"
+#include "classfile/systemDictionaryShared.hpp"
+#include "classfile/vmClasses.hpp"
+#include "compiler/compilationPolicy.hpp"
+#include "gc/shared/gcVMOperations.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/constantPool.inline.hpp"
+#include "oops/instanceKlass.hpp"
+#include "oops/klass.inline.hpp"
+#include "runtime/handles.inline.hpp"
+#include "runtime/javaCalls.hpp"
+#include "runtime/perfData.inline.hpp"
+#include "runtime/timer.hpp"
+#include "services/management.hpp"
+
+void AOTLinkedClassBulkLoader::serialize(SerializeClosure* soc, bool is_static_archive) {
+  AOTLinkedClassTable::get(is_static_archive)->serialize(soc);
+}
+
+void AOTLinkedClassBulkLoader::load_javabase_boot_classes(JavaThread* current) {
+  load_classes_in_loader(current, LoaderKind::BOOT, nullptr);
+}
+
+void AOTLinkedClassBulkLoader::load_non_javabase_boot_classes(JavaThread* current) {
+  load_classes_in_loader(current, LoaderKind::BOOT2, nullptr);
+}
+
+void AOTLinkedClassBulkLoader::load_platform_classes(JavaThread* current) {
+  load_classes_in_loader(current, LoaderKind::PLATFORM, SystemDictionary::java_platform_loader());
+}
+
+void AOTLinkedClassBulkLoader::load_app_classes(JavaThread* current) {
+  load_classes_in_loader(current, LoaderKind::APP, SystemDictionary::java_system_loader());
+}
+
+void AOTLinkedClassBulkLoader::load_classes_in_loader(JavaThread* current, LoaderKind loader_kind, oop class_loader_oop) {
+  if (!CDSConfig::is_using_aot_linked_classes()) {
+    return;
+  }
+
+  HandleMark hm(current);
+  ResourceMark rm(current);
+  ExceptionMark em(current);
+
+  Handle h_loader(current, class_loader_oop);
+
+  load_table(AOTLinkedClassTable::for_static_archive(),  loader_kind, h_loader, current);
+  assert(!current->has_pending_exception(), "VM should have exited due to ExceptionMark");
```

Comment:
An `ExceptionMark` only triggers a VM exit on construction and destruction, if an exception is pending.

Reply:
I refactored the code a bit to handle any exception that occurs during bulk loading, and print our more useful error messages.